### ISSUE TITLE
Problem 1.3 URLify

### ISSUE THREAD

**Issue by rando3** [careercup/CtCI-6th-Edition-Python#70]
Solution assumes type str is mutable in python when it is obviously immutable. How can this be the solution when it tries to mutate a string?

**Comment by Twoody** [comment]
@rando3 please elaborate. If I have 
    >>>s = "foo"`
    >>>s[0]= "b" 
    >>>print(s) 
   \>\>\>boo

What is immutable about this?

**Comment by rando3** [comment]
Don’t modify strings! Strings in python (and various languages) are immutable for many reasons. You will get a type error for string assignment. You can’t change them in place, but you can create a new string or work with strings as lists!

**Comment by rando3** [comment]
This is easily StackOverflow-able if you don’t understand what I’m talking about. https://stackoverflow.com/questions/1228299/change-one-character-in-a-string

**Comment by Twoody** [comment]
@rando3 yes, but the string is mutable. Moreover, the function calls list on the items and treats the string as a list (much like the StringBuilder class in java which the book asks you to use). Are you asking to change the variable name to `stringAsList`?

**Comment by lrogar** [comment]
Maybe something like this? As an alternative to having to convert the array into a list. This is much faster.

```python
def urlify(string, length):

    string = string.strip()
    curr_index = 0

    for char in string:
        if char == ' ':
            string = string[:curr_index] + "%20" + string[curr_index+1:]
            curr_index+=3
            continue
        curr_index+=1
    return string
```

**Comment by priyankasrs** [comment]
This takes even lesser time.
`def urilify(string):
    return string.strip().replace(" ", "%20")`

**Comment by ianbrayoni** [comment]
@rando3 You are onto something, I also got tripped by the idea that the solution appears to treat strings as `mutable` when in fact strings are not. Calling the function as follows:

```
urlify('Mr John Smith    ', 13)
```

will definetly run into `TypeError: 'str' object does not support item assignment` due to the assignments on either **line 12** or **line 16** in the solution [URLify.py](https://github.com/careercup/CtCI-6th-Edition-Python/blob/master/Chapter1/3_URLify/URLify.py)


 However on examining the test cases:

```
(list('Mr John Smith    '), 13, list('Mr%20John%20Smith'))
```
`list` is called on the argument string and the expected return value which makes the function interface all the more confusing 👎 . 

Nevertheless, using `list()` ensures that the processing within the function is done on a `mutable` data structure `list` hence @Twoody's point.

Having said that,
> I would recommend that the function be rectified to take a `string` and an `int` as argument types to prevent confusion as to what types one needs to pass to it. 

In fact, that is what the question implies in its example.

```
EXAMPLE
Input: "Mr John Smith    ", 13
Output: "Mr%20John%20Smith"
```

A solution that would clear this up bearing in mind the assumption stated in the question. I added the 4 spaces to clarify _sufficient_.

> You may assume that the string has sufficient space at the end (4 spaces) to hold the additional characters, and that you are given the "true" length of the string.

```
def urlify(string, trueLength):
    index = len(string)
    stringL = list(string)

    for i in reversed(range(trueLength)):
        if stringL[i] == " ":
            # Replace spaces
            stringL[index - 3:index] = "%20"
            index -= 3
        else:
            # Move characters
            stringL[index - 1] = stringL[i]
            index -= 1

    return "".join(stringL)
```

This would take `O(N)` _time_ because we are scanning and updating the list and `O(N)` _space_ because we store the string in a list so as to manipulate it.

@luisarojas Great solution! However a concern an interviewer would raise is that you did not use some of the details provided in the question e.g:
1. "true" length of the string.
(which is passed to your function but not used - this can be treated as a code smell too)
2. The assumption that the string has sufficient space at the end to hold the
additional characters.

@priyankasrs your solution is great and solves the problem in one pass but I am not sure if an interviewer would take it because in an interview its preferrable if you can break down and solve the problem while explaining your thought process in the 30 - 45 minutes slot allocated to you. One rule of thumb I use is that if the solution is so succinct and relies heavily on the standard lib...its probably not what the interviewer wants.
Besides, you are not using the second argument at all - "true" length of the string.

P.S: [MyPy](https://github.com/python/mypy)

**Comment by riazanovslv** [comment]
What about this? Utilizing both parameters and not using intermediate lists
```
def urlify(string, length):
    return ''.join('%20' if c == ' ' else c for c in string[:length])
```

**Comment by aeroreyna** [comment]
@ianbrayoni
Even if the problem statement says:

> You may assume that the string has sufficient space at the end (no of spaces * 2) to hold the additional characters, and that you are given the "true" length of the string.

This doesn't meant it will have just the right space for it, it might be longer than required.
Doesn't it?

This is my implementation, using list to transform the string into an array-like data structure.

~~~python
def replace_space(url, l):
    count = 0
    for i in range(l):
        count += url[i] == " "
    for i in range(l):
        j = l-i-1
        if url[j] != " ":
            url[j + count*2] = url[j]
        else:
            url[j + count*2] = "0"
            url[j + count*2-1] = "2"
            url[j + count*2-2] = "%"
            count -= 1
    return url

if __name__ == '__main__':
    print(replace_space(list("I am I               "), 6))
~~~

**Comment by abnerabbey** [comment]
**My solution in Python with _Recursion_**

`def urlify(s):
    if len(s) <= 1 and s != '':
        return s
    if s[0] == ' ' and s[1] == ' ':
        return '' + urlify(s[1:])
    elif s[0] == ' ' and s[1] != ' ':
        return '%20' + urlify(s[1:])
    return s[0] + urlify(s[1:])

print(urlify('Mr       John          Smith    '))`

**Comment by abnerabbey** [comment]
` def urlify(s):
    if len(s) <= 1 and s != '':
        return s
    if s[0] == ' ' and s[1] == ' ':
        return '' + urlify(s[1:])
    elif s[0] == ' ' and s[1] != ' ':
        return '%20' + urlify(s[1:])
    return s[0] + urlify(s[1:])

print(urlify('Mr       John          Smith    ')) `

**Comment by abnerabbey** [comment]
`
let a = 13
`

**Comment by abnerabbey** [comment]
**My solution with _Recursion_**
`
def urlify(s):
if len(s) <= 1 and s != '':
return s
if s[0] == ' ' and s[1] == ' ':
return '' + urlify(s[1:])
elif s[0] == ' ' and s[1] != ' ':
return '%20' + urlify(s[1:])
return s[0] + urlify(s[1:])

print(urlify('Mr John Smith '))
`

**Comment by abnerabbey** [comment]
`
def urlify(s):
    if len(s) <= 1 and s != '':
        return s
    if s[0] == ' ' and s[1] == ' ':
        return '' + urlify(s[1:])
    elif s[0] == ' ' and s[1] != ' ':
        return '%20' + urlify(s[1:])
    return s[0] + urlify(s[1:])
`

**Comment by brycedrennan** [comment]
I believe the original issue has been resolved now by #124.

**Comment by archie-bryann** [comment]
> This takes even lesser time. `def urilify(string): return string.strip().replace(" ", "%20")`

Exactly.
